### DISCUSSION TITLE
system_open command cannot open certain types of files properly

### DISCUSSION THREAD

**Discussion by la1den** [nvim-tree/nvim-tree.lua#2528] [Q&A] [answered]
### Description

I use nvim-tree's system open on the Linux desktop, but some types of files cannot be opened, while others can.

### Neovim version

```text
NVIM v0.9.2
Build type: Release
LuaJIT 2.1.1694285958
```

### Operating system and version

Linux 6.1.56-1-lts x86_64 rolling n/a

### nvim-tree version

16a0e3c

### Clean room replication

```lua
-- disable netrw at the very start of your init.lua (strongly advised)
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1
-- set termguicolors to enable highlight groups
vim.opt.termguicolors = true
-- empty setup using defaults
require("nvim-tree").setup()
```

### Steps to reproduce

1. Use nvim-tree to open a directory with a js suffix for a file
2. Press s to attempt to open the JS file

**Comment by la1den** [comment]
I think it may be related to input and output, so I tried to disable output redirection and it worked properly, but I don't know why.

https://github.com/nvim-tree/nvim-tree.lua/assets/55609004/9bb84ae9-c0ef-4c26-b198-a0fdedd135a8

**Comment by alex-courtis** [comment, marked as answer]
That's... just how `xdg-open` works. nvim-tree will simply call `system_open.cmd` with the absolute file path as the argument.

Run `xdg-open something.js` directly from a terminal to see what is going on.

Long ago I overrode xdg-open and have never looked back:
```sh
: ; cat /usr/local/bin/xdg-open
#!/bin/sh

# Suppress any spurious output from noisy apps like chrome.
# Invoke /usr/bin/xdg-open directly for more interactive operations.

/usr/bin/xdg-open "${@}" > /dev/null 2>&1
```

**Reply by la1den** [reply]
https://github.com/nvim-tree/nvim-tree.lua/blob/874ae6e9445a5eb5ba430e5fd10212450a261ad7/lua/nvim-tree/actions/node/system-open.lua#L22
`stdio = process.cmd == 'xdg-open' and {nil, nil, nil} or { nil, nil, process.stderr }`
Compared to modifying the system's xdg-open command, is it better to determine whether the cmd is xdg-open in the system-open file ?If other applications of the system use xdg-open, will they be affected by the modified xdg-open?

**Reply by alex-courtis** [reply]
Safest way might be a wrapper script on the path instead of overriding. You could explicitly set that e.g.

```lua
system_open.cmd = "my-quiet-xdg-open"
```

**Comment by ensorpalacios** [comment]
Hi @alex-courtis 

I think I have a related problem. On ubuntu I've changed default pdf viewer from Document Viewer to Okular, and now I cannot open pdf's anymore with system open. I've tried to change to xdg-open but again nothing happens when pressing `s`. Do you have any advise? This is my nvim-tree.lua

```lua
-- this is a local function
local function my_on_attach(bufnr)
  local api = require 'nvim-tree.api'

  local function opts(desc)
    return { desc = 'nvim-tree: ' .. desc, buffer = bufnr, noremap = true, silent = true, nowait = true }
  end

  -- default mappings
  api.config.mappings.default_on_attach(bufnr)

  -- custom mappings
  vim.keymap.set('n', 'u', api.tree.change_root_to_parent, opts 'Up') -- changed from default <C-t>
  vim.keymap.set('n', '?', api.tree.toggle_help, opts 'Help')
end

-- pass to setup along with your other options
return {
  'nvim-tree/nvim-tree.lua',
  version = '*',
  lazy = false,
  dependencies = {
    'nvim-tree/nvim-web-devicons',
  },
  config = function()
    require('nvim-tree').setup {
      vim.keymap.set('n', '<leader>st', '<cmd>NvimTreeOpen<cr>'),
      sort = {
        sorter = 'case_sensitive',
      },
      view = {
        width = 30,
      },
      renderer = {
        group_empty = true,
      },
      filters = {
        dotfiles = true,
      },
      on_attach = my_on_attach,
      system_open = {
        cmd = '/usr/bin/xdg-open',
      },
    }
  end,
}
```

Many thanks for the support

**Reply by gegoune** [reply]
Does `/usr/bin/xdg-open <path/file>` work?

**Reply by ensorpalacios** [reply]
hi @gegoune

It gives me the following warning, but yes, I can open a pdf with that.

```
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
Icon theme "breeze" not found.
```

**Reply by alex-courtis** [reply]
We are defaulting to the neovim native vim.ui.open as of [1.7.0](https://github.com/nvim-tree/nvim-tree.lua/releases/tag/nvim-tree-v1.7.0)

It will be used when `system.open.cmd` is not specified and is preferred, as it is more robust. Please remove `system.open.cmd`.

You can also test this via `:lua vim.ui.open('foo.pdf')`

**Reply by ensorpalacios** [reply]
@alex-courtis sorry but :lua vim.ui.open('foo.pdf') doesn't work either when setting okular as default (with Document viewer it does work)

**Reply by alex-courtis** [reply]
`:lua print(vim.inspect(vim.ui.open('foo.pdf')))` doesn't provide any information beyond it's successfully starting the process.

Did you test with an actual pdf file? xdg won't open a pdf file that doesn't exist or is empty.

I'm afraid we have left the scope of nvim-tree, and I suspect it's an issue with your xdg-mime configuration which is not under neovim's control. You could raise this in a neovim forum.

**Reply by ensorpalacios** [reply]
No problem thank you anyway for the help
